src/invessel_build.py

```diff
@@ -0,0 +1,480 @@
+            self,
+            vmec,
+            build,
+            repeat,
+            num_phi=61,
+            num_theta=61,
+            scale=m2cm,
+            plasma_h5m_tag=None,
+            sol_h5m_tag=None,
+            logger=None
+    ):
+        self.vmec = vmec
+        self.build = build
+        self.repeat = repeat
+        self.num_phi = num_phi
+        self.num_theta = num_theta
+        self.scale = scale
+        self.plasma_h5m_tag = plasma_h5m_tag
+        self.sol_h5m_tag = sol_h5m_tag
+
+        self.logger = logger
+        if self.logger == None or not self.logger.hasHandlers():
+            self.logger = log.init()
+
+        try:
+            assert (self.repeat + 1) * self.build['phi_list'][-1] <= 360.0, (
+                'Total toroidal extent requested with repeated geometry '
+                'exceeds 360 degrees. Please examine phi_list and the repeat '
+                'parameter.'
+            )
+        except AssertionError as e:
+            self.logger.error(e.args[0])
+            raise e
+
+        if self.build['wall_s'] != 1.0:
+            self.build['radial_build'] = {
+                'sol': {
+                    'thickness_matrix': np.zeros((
+                        len(self.build['phi_list']),
+                        len(self.build['theta_list'])
+                    )),
+                    'h5m_tag': self.sol_h5m_tag
+                },
+                **self.build['radial_build']
+            }
+
+        self.build['radial_build'] = {
+            'plasma': {
+                'thickness_matrix': np.zeros((
+                    len(self.build['phi_list']),
+                    len(self.build['theta_list'])
+                )),
+                'h5m_tag': self.plasma_h5m_tag
+            },
+            **self.build['radial_build']
+        }
+
+        self.phi_list = expand_ang(self.build['phi_list'], self.num_phi)
+        self.theta_list = expand_ang(self.build['theta_list'], self.num_theta)
+
+    def populate_surfaces(self):
+        '''Populates Surface class objects representing the outer surface of
+        each component specified in the radial build.
+        '''
+        self.Surfaces = []
+
+        offset_mat = np.zeros((
+            len(self.build['phi_list']), len(self.build['theta_list'])
+        ))
+
+        for name, layer_data in self.build['radial_build'].items():
+            try:
+                assert np.all(np.array(layer_data['thickness_matrix']) >= 0), (
+                    'Component thicknesses must be greater than or equal to 0. '
+                    'Check thickness inputs for negative values.'
+                )
+            except AssertionError as e:
+                self.logger.error(e.args[0])
+                raise e
+
+            if name == 'plasma':
+                s = 1.0
+            else:
+                s = self.build['wall_s']
+
+            if 'h5m_tag' not in layer_data:
+                layer_data['h5m_tag'] = name
```

Comment:
I think the above logic of the defaults means that the `h5m_tag` for plasma and sol could be `Null`. Is that OK? Or do we need to test inputs to prevent that.

Reply:
Good point. The intended workflow is that if `plasma_h5m_tag` and `sol_h5m_tag` are `None`, the applied material tag for each will be set to the respective name (i.e., the tag for the plasma will be `'plasma'` and that for the SOL will be `'sol'`) but it seems I accidentally dropped that functionality during my change to OO. I'll rework the script to make sure that functionality is restored, unless you have a suggestion for an alternative workflow.